File: 2024/solutions/day02.go
package solutions

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

type Day02 struct{}
type input02 struct {
	Reports [][]int
}

func (d Day02) Parse(raw string) input02 {
	lines := strings.Split(raw, "\n")
	reports := make([][]int, len(lines))
	for i, line := range lines {
		nums := make([]int, 0)
		for _, num := range strings.Split(line, " ") {
			n, _ := strconv.Atoi(num)
			nums = append(nums, n)
		}
		reports[i] = nums
	}
	return input02{reports}
}

func (d Day02) Part1(raw string) {
	fmt.Println("Day 2 Part 1")
	input := d.Parse(raw)

	count := 0
	for _, report := range input.Reports {
		if safe(report) {
			count++
		}
	}

	fmt.Printf("Solution: %d\n", count)
}

func (d Day02) Part2(raw string) {
	fmt.Println("Day 2 Part 2")
	input := d.Parse(raw)

	count := 0
	for _, report := range input.Reports {
		if safe(report[1:]) || safe(report[:len(report)-1]) {
			count += 1
			continue
		}
		for i := 1; i+1 < len(report); i++ {
			short := make([]int, 0)
			short = append(short, report[:i]...)
			short = append(short, report[i+1:]...)
			if safe(short) {
				count += 1
				break
			}
		}
	}

	fmt.Printf("Solution: %d\n", count)
}

func safe(report []int) bool {
	last := report[0]
	inc := report[0] < report[1]
	for i := 1; i < len(report); i++ {
		n := report[i]
		if last == n || last < n != inc || 3 < math.Abs(float64(n-last)) {
			return false
		}
		last = n
	}
	return true
}

File: rust/2024/src/solutions/day_02.rs
use pathfinding::num_traits::abs;

use super::day_trait::Day;

type Report = Vec<usize>;

#[derive(Debug)]
pub struct Input {
    lines: Vec<String>,
    reports: Vec<Report>,
}

fn parse_input(raw: &str) -> Input {
    let lines: Vec<String> = raw.split("\n").map(|s| s.to_owned()).collect();
    let reports = lines
        .iter()
        .map(|l| l.split(" ").map(|d| d.parse().unwrap()).collect())
        .collect();
    Input { lines, reports }
}

pub struct Day02;
impl Day for Day02 {
    fn day(&self) -> u8 {
        02
    }

    fn part_1(&self, raw: &str) {
        println!("Day {} part 1", self.day());
        let input: Input = parse_input(raw);

        let count: usize = input.reports.iter().filter(|&r| safe(r)).count();

        println!("Answer is {}", count);
    }

    fn part_2(&self, raw: &str) {
        println!("Day {} part 2", self.day());
        let input: Input = parse_input(raw);

        let count: usize = input
            .reports
            .iter()
            .filter(|&r| {
                (0..r.len()).any(|i| {
                    let mut q = r.clone();
                    q.remove(i);
                    safe(&q)
                })
            })
            .count();

        println!("Answer is {}", count);
    }
}

fn safe(report: &Report) -> bool {
    let increasing = report[0] < report[1];
    report.windows(2).all(|w| {
        w[0] != w[1] && abs(w[0] as f32 - w[1] as f32) <= 3. && (w[0] < w[1]) == increasing
    })
}
